tracing-journald/src/socket.rs

```diff
@@ -0,0 +1,62 @@
+//! socket helpers.
+
+use std::io::{Error, Result};
+use std::mem::size_of;
+use std::os::unix::net::UnixDatagram;
+use std::os::unix::prelude::{AsRawFd, RawFd};
+use std::ptr;
+
+use libc::*;
+
+const CMSG_BUFSIZE: usize = 88;
```

Comment:
@Ralith @hawkw I've tried to implement the cmsg stuff along what I've seen in quinn-rs (as recommended by @Ralith ), and the code examples in `cmsg(3)` and related documentation, and I'm seeing something that I don't quite understand:

If I pick a buffer size of 39 or less the fd descriptor makes it to journald and I see the message appearing in the journal, and the new test passes.

But if I pick a buffer size of 40 or higher `sendmsg` returns `EINVAL`.

I don't really understand what's going on, and I find this quite hard to debug :thinking: 

Do you perhaps have any idea or any pointers?  Or is there an obvious mistake in my code?

Reply:
hmm, that's very strange. i spent some time looking at the `cmsg` and `sendmsg` documentation, and nothing really stood out to me, i'm not sure what's going on here. i'm not particularly familiar with Unix socket APIs, so perhaps @Ralith knows?

Reply:
The bug is probably that you're setting `msg_controllen = CMSG_BUFSIZE`, when `CMSG_BUFSIZE` is the size of the *buffer*, but `msg_controllen` expects the size of the *message*. You should use `CMSG_SPACE` to compute the correct value for `msg_controllen` for the target platform.

Reply:
@Ralith Indeed, this works, and `cmsg(3)` actually says that:

![image](https://user-images.githubusercontent.com/224922/143501032-a2bc8c43-4bfa-4cd5-a82e-97e57eba0417.png)

I was confused by the documentation of the `msghdr` struct in `sys/socket.h`:

![image](https://user-images.githubusercontent.com/224922/143501235-32502106-84c2-40d7-802b-b2e761da41a9.png)